ArduCopter/precision_landing.cpp

```diff
@@ -22,4 +27,197 @@ void Copter::update_precland()
+            // If not, we will retry landing
+            landing_target_lost_action = Landing_Target_Lost_Action::DESCEND;
+        } else if (strictness == AC_PrecLand::PldRetryStrictness::NOT_STRICT) {
+            // User just wants to land, prec land isn't a priority
+            landing_target_lost_action = Landing_Target_Lost_Action::LAND_VERTICALLY;
+        }
+    }
+        break;
+
+    case Landing_Target_Lost_Action::DESCEND:
+        copter.flightmode->run_land_controller();
+        if (AP_HAL::millis() - copter.precland.get_last_precland_output_ms() >= copter.precland.get_min_retry_time_sec()) {
+            // we have descended for sometime and the target still isn't in sight
+            // lets retry
+            landing_target_lost_action = Landing_Target_Lost_Action::RETRY_LANDING;
+            retry_status = RetryLanding::RETRY_INIT;
+        }
+        break;
+
+    case Landing_Target_Lost_Action::RETRY_LANDING:
+        // retry the landing by going to the last known horizontal position of the target
+        retry_landing();
+        break;
+
+    case Landing_Target_Lost_Action::LAND_VERTICALLY:
+    default:
+        // Just land vertically
+        copter.flightmode->run_land_controller();
+        break;
+    }
+}
+
+// Retry landing based on a previously known location of the landing target
+void Mode::PrecLand_StateMachine::retry_landing()
+{
+    if (_retry_count > copter.precland.get_max_retry_allowed() || copter.precland.get_max_retry_allowed() == 0) {
+        // we have exhausted the amount of times vehicle was allowed to retry landing
+        // do failsafe measure so the vehicle isn't stuck in a constant loop
+        do_failsafe_actions();
+        return;
+    }
+
+    // get the last known location of the landing target
+    Vector3f last_known_loc;
+    copter.precland.get_last_detected_landing_location(last_known_loc);
+
+    switch (retry_status)
+    {
+    case RetryLanding::RETRY_INIT:
+        // Init the Retry
+        _retry_count ++;
+        copter.flightmode->land_retry_position(last_known_loc);
+        retry_status = RetryLanding::RETRY_IN_PROGRESS;
+        // inform the user what we are doing
+        copter.gcs().send_text(MAV_SEVERITY_INFO, "Retrying Precision Landing");
+        break;
+
+    case RetryLanding::RETRY_IN_PROGRESS:
+    {
+        // continue converging towards the target till we are close by
+        copter.flightmode->land_retry_position(last_known_loc);
+        const float dist_to_target_xy = copter.pos_control->get_pos_error_xy_cm();
+        const float dist_to_target_z = copter.pos_control->get_pos_error_z_cm();
+        if ((dist_to_target_xy < MAX_POS_ERROR_CM) && (fabsf(dist_to_target_z) < MAX_POS_ERROR_CM)) {
+            // we have approx reached landing location previously detected
+            retry_status = RetryLanding::RETRY_COMPLETE;
+            copter.gcs().send_text(MAV_SEVERITY_INFO, "Landing Retry Completed");
```

Comment:
Slight problem with the message in that you'd consider your "landing try" completed if you actually landed, so by extension your "landing retry" would be complete if you actually landed.  This is more "landing retry reset complete"

Would be interesting to lint our code for tense here - statement of current state or statement of accomplishment.